OCI Routes

argha-dot opened the pull request:
- moved api layers to it's own api folder, 
 - ~created a registry data store connection layer~
 - ~added put, post, and patch request for blob upload~
 - ~add get request for tag list~
 - ~add manifest upload request~
 - TDD -> cleanup while testing
 - Etags everywhere
 - Response builder API everywhere
 - Fix streaming in PATCH blob and PULL blob
 - cargo prepare
 - fix sizes
 - migrate to aws_s3_sdk
 
 
- [x] **end-1 — GET /v2/**
  - [x] `GET /v2/` → `200`

- [x] **end-2 — /v2/<name>/blobs/<digest>**
  - [x] `GET /v2/<name>/blobs/<digest>` → `200`
  - [x] `HEAD /v2/<name>/blobs/<digest>` → `200`

- [x] **end-3 — /v2/<name>/manifests/<reference>**
  - [x] `GET /v2/<name>/manifests/<reference>` → `200`
  - [x] `HEAD /v2/<name>/manifests/<reference>` → `200`

- [x] **end-4a — POST /v2/<name>/blobs/uploads/**
- [x] **end-4b — POST /v2/<name>/blobs/uploads/?digest=<digest>**
- [x] **end-5 — PATCH /v2/<name>/blobs/uploads/<reference>**
- [x] **end-6 — PUT /v2/<name>/blobs/uploads/<reference>?digest=<digest>**
- [x] **end-7 — PUT /v2/<name>/manifests/<reference>**

- [x] **end-8a — GET /v2/<name>/tags/list**
- [ ] **end-8b — GET /v2/<name>/tags/list?n=<integer>&last=<tagname>**

- [ ] **end-9 — DELETE /v2/<name>/manifests/<reference>**
- [ ] **end-10 — DELETE /v2/<name>/blobs/<digest>**

- [ ] ~**end-11 — POST /v2/<name>/blobs/uploads/?mount=<digest>&from=<other_name>**~

- [ ] ~**end-12a — GET /v2/<name>/referrers/<digest>**~
- [ ] ~**end-12b — GET /v2/<name>/referrers/<digest>?artifactType=<artifactType>**~

- [ ] **end-13 — GET /v2/<name>/blobs/uploads/<reference>**

argha-dot committed:
moved api layers to it's own api folder, and created a registry data store connection layer

argha-dot committed:
add upload route

argha-dot committed:
add manifest upload route

argha-dot commented:
TODOS
 - [ ] delete the session storage blob
 - [ ] Verify Content Lengths from body
 - [x] Tags Table
 - [x] PUT and PATCH methods for upload [blob](https://github.com/opencontainers/distribution-spec/blob/main/spec.md#post-then-put)
 - [x] Adding case for POST upload blob when the digest isn't given i.e. creating the session id 
 - [x] Creating a session table
 - [x] Change DB schema for manifest to support layes with annotations

rakshith-ravi commented on api/src/routes/registry.patr.cloud/blobs/uploads.rs:
```diff
@@ -0,0 +1,172 @@
+
+	let digest = query.digest;
+	if digest.is_empty() {
+		todo!();
```
```suggestion
		todo!("create a session ID if a digest doesn't exist");
```

rakshith-ravi commented on api/src/routes/registry.patr.cloud/blobs/uploads.rs:
```diff
@@ -0,0 +1,172 @@
+		.await
+		.map_err(internal_server_error_response)?;
+
+	// if !digest.to_string() {
+	// 	todo!("Create this later")
+	// }
```
Remove

rakshith-ravi commented on api/src/routes/registry.patr.cloud/manifest/push.rs:
```diff
@@ -0,0 +1,133 @@
+		));
+	}
+
+	// TODO: This also needs to be completed, but rn I don't have the right
+	// column
+	// query!(
+	//     r#"
+	//         INSERT INTO
+	//     "#
+	// )
+	// .execute(&mut *database)
+	// .await
+	// .map_err(internal_server_error_response)?;
```
What is this?

rakshith-ravi commented on api/src/routes/registry.patr.cloud/mod.rs:
```diff
@@ -29,20 +29,31 @@ pub async fn setup_routes(state: &AppState) -> Router {
 				),
 		)
-		.with_state(state.clone())
+		.with_state(())
```
This is useless

rakshith-ravi commented on api/src/routes/registry.patr.cloud/mod.rs:
```diff
@@ -29,20 +29,31 @@ pub async fn setup_routes(state: &AppState) -> Router {
+	format!("registry/manifest/{manifest}")
+}
+
+// .route(
+// 	"/:workspaceId/:repoName/blobs/:digest",
+// 	get(get_blob_info::handle).head(get_blob_info::handle),
+// )
+// .route(
+// 	"/:workspaceId/:repoName/manifests/:reference",
+// 	get(get_manifest_info::handle),
+// ),
```
Remove

argha-dot committed:
add tags list route, add put method for upload blob, and session creation in POST upload blob

argha-dot committed:
add patch route for blob upload

Co-authored-by: Rakshith Ravi <rakshith-ravi@users.noreply.github.com>

rakshith-ravi committed:
add registry listener and update routes for blob and manifest handling

argha-dot committed:
add get/head routes for manifest/reference

argha-dot committed:
remove container_registry_index and finish PUT

argha-dot committed:
fix route errors

argha-dot committed:
test head blob, post manifest route

rakshith-ravi committed:
Finally! A blob that partially comes on S3.

Added declare registry route as well

argha-dot committed:
working push errored pull

argha-dot committed:
fix failed session uploads error

rakshith-ravi commented:
Rebased for the sake of the merge conflicts

rakshith-ravi committed:
Updated sqlx files

argha-dot committed:
working push and pull

argha-dot committed:
add todo for disabling single POST

argha-dot committed:
delete route for manifest and delete not allowed for blob

rakshith-ravi committed:
Update dependencies and refactor registry endpoint handling

rakshith-ravi committed:
feat: Implement endpoint handling and request preprocessing layers
Remove obsolete registry layers: `data_store_connection_handler.rs` and `request_parser.rs`.

rakshith-ravi committed:
What the fuck.

So many changes. How am I supposed to describe this in the commit message.

I guess this is a partial rewrite of the registry. So many things cleaned up. So much more cleanup to do. I fucking hate this. My only solace is that once this is implemented I won't have to touch this again. Fuck this

rakshith-ravi committed:
Refactor for AWS SDK migration: updated Cargo.toml, replaced rust-s3 with aws-sdk-s3 across handlers, revamped error handling, added ETag/Cache-Control, updated S3 client initialization, and introduced a TotalCountHeader utility.

rakshith-ravi committed:
refactor: Update manifest handlers to improve clarity and functionality; adjust delete endpoint comments and modify repository name handling in upload manifest

rakshith-ravi committed:
- All registry endpoints are now authenticated. You will be authenticated and you WILL be okay it.
- Removed the unused `auth_endpoint_handler` module from the registry layer.
- The entirity of the registry is now authenticated, but the token isn't being passed to the subsequent layers anymore (since I don't feel like it's necessary).
- Added regex constants for validating registry repository names, tags, and digests.
- Introduced new headers for Docker upload UUID and Location, enhancing header management.
- Added a new Docker login endpoint for user authentication.
- All S3 endpoints are now streamed.
- Registry routes are now using AWS S3 SDK.

TODO:
- Uploading a chunk for a blob currently doesn't work, but that's an upstream issue. Can consider streaming to a file and putting that file upstream.
- When a blob is uploaded, but before a manifest is, a HEAD request is made by the client to check if the blob got uploaded correctly. This currently fails because of a permission check. Need to figure out a way around that.

rakshith-ravi committed:
refactor: Remove unused S3Error conversion from RegistryError

rakshith-ravi committed:
refactor: Update SQL queries in deployment handlers to use container_registry_tag and improve clarity
